MFLT-4887 Document new metric traits and limits (#127)
memfault/memfault-docs: docs/platform/metrics.mdx

---

> title: Metrics

Reviewer: For scalability of the side bar menu, should we call this "Metrics and Attributes"? I am worried that readers otherwise will have a hard time to read up on "Device Attributes".

Author: Agreed, I have my reservations about the side bar as it is, I think it's unusable unless you know specifically what you're looking for. We need some sort of ordering and more nesting, I think. The menu doesn't reflect our product offerings at all.

---

import ProjectSettingsMetricsPath from "@site/src/components/ProjectSettingsMetricsPath";

Metrics are a way to proactively monitor any smart device. You can monitor your
releases and view data-driven device and fleet-level Metrics, like battery
health, connectivity, and more using Memfault's Metrics features:
[Metric Charts](/docs/platform/metric-charts),
[Device Sets](/docs/platform/metrics#device-sets),
[Alerts](/docs/platform/alerts). Metrics also power parts of our Device view

> such as Device Attributes and the Timeline.

Reviewer: "Metrics also power parts…" – I'd pull up that sentence to become the second of this paragraph.

Metrics can be marked as Timeseries as well as Attributes, and each option
enables a different set of features in the Memfault app.

## Defining a Metric

> Metrics are created automatically when devices report them from the field. Here

Reviewer: Docs should also mention that one can create them manually in the UI

are some useful links if you're looking to enable Metrics collection from your
devices using our SDKs:

- for MCU: [Collecting Device Metrics using our MCU SDK](/docs/mcu/metrics-api)
- for Android:
  [Collecting Device Metrics using our Android SDK](/docs/android/android-eventlog-metrics).

### From the Memfault app

You can view all defined Metrics by going to <ProjectSettingsMetricsPath />. A
Metric can have a few properties:

- a `string_key` which devices and APIs can use to identify the Metric
- whether the data type is an `INT` (default), `FLOAT`, `BOOL`, or `STRING`
- whether it is a `favorite` Metric, which will cause it to appear more
  prominently in a few places in the web app
- the value `range` which informs charts and other components
- a set of `ignored_values` (like `-1`) which should be ignored.

## Timeseries

By marking a Metric as a Timeseries you'll be able to use it in
[Device Sets](/docs/platform/metrics#device-sets),
[Metric Charts](/docs/platform/metric-charts), and

> [Alerts](/docs/platform/alerts). You'll also be able to search by it with a time

Reviewer: ```suggestion
[Alerts](/docs/platform/alerts). You'll also be able to use them to search for devices  with a time
```

constraint (e.g. `battery_level > 90 anytime last week`).

To mark a Metric as a Timeseries, go to <ProjectSettingsMetricsPath /> and click
**Edit** on the Metric you're interested in. Then check the **Timeseries**
checkbox.

## Attributes

> You can store custom Attributes for each of your devices.

Reviewer: …to build a device registry or to manage inventory.

By marking a Metric as an Attribute it will be shown in the Attributes tab on
device pages. You'll also be able to search for devices by Attributes it on the
device list page and define Device Sets.

### Defining an Attribute

In <ProjectSettingsMetricsPath /> you'll be able to mark specific Metrics as
Attributes. Metrics reported by devices are automatically marked as
Attributes[<sup>\*</sup>](#quotas). You can configure this by clicking **Edit**
on a Metric you're interested in, or by creating a new Metric with the Attribute
feature enabled.

#### Device Write Policy

Attributes have a `device_write_policy` property that controls under which
condition a device can change its value. This does not affect the ability to set
a value manually on the device page.

<table>
  <tbody>
    <tr>
      <td>Writable</td>
      <td>
        Devices are always allowed to change the value of this Attribute. This

>         is the default. Example: Location of a mobile device that changes

Reviewer: can we better visually set apart the example here and the rows below?

Author: ![image](https://user-images.githubusercontent.com/16181067/142196854-137166f4-a1cf-445b-b37d-d459ebf43864.png)

:::note

Because devices can update Attributes, anyone who holds your **Project Key** can
write or overwrite these Attributes.

:::

#### Using the REST API

Below is an example of how Attributes can be set using the Project Key and the
REST API:

:::note

Be sure to replace `${DEVICE_SERIAL}` and `${YOUR_PROJECT_KEY}` below with the
device you want to update and your
[Project Key](/docs/ci/authentication#project-key), respectively.

:::

```bash
curl --location --request PATCH \
"https://device.memfault.com/api/v0/attributes?device_serial=${DEVICE_SERIAL}" \
--header "Content-Type: application/json" \
--header "Memfault-Project-Key: ${YOUR_PROJECT_KEY}" \
--data-raw '
[
  {"string_key": "ProductName", "value": "SmartWatch"},
  {"string_key": "ModelColor", "value": "Blue"}
]
'
```

#### Using the Web Application

Visit the respective device page as described in
[Viewing an Attribute](#viewing-an-attribute). Click on the _pencil icon_ next
to the value, enter the new value, and finally **Save** it.

<p align="center">
  <img
    src="/img/blog/2021-08-17-device-properties-edit.png"
    alt="Example of editing an Attribute via the web application"
  />
</p>

> ### Searching by Attribute

Reviewer: ```suggestion
### Searching Devices by Attribute
```

Reviewer: Also, should we have the equivalent for Timeseries Metrics?

You can search devices by Attributes. Visiting the **Devices** page, open the
**Filter** panel and select the Metric you want to search by.

<p align="center">
  <img
    src="/img/blog/2021-08-17-device-properties-filter.png"
    alt="Example of filtering by an Attribute"
  />
</p>

You can search by multiple Attributes using the _plus icon_.

:::note Searching by Attribute vs Timeseries

If you select a **During** time frame, devices will be searched by their
Timeseries Metrics during that time frame, not by their device Attributes.

:::

> ## Quotas

Reviewer: should there be sentence about rate limiting + link to the respective item?
